ray-operator/controllers/ray/raycluster_controller_test.go

```diff
@@ -255,6 +255,38 @@ var _ = Context("Inside the default namespace", func() {
 				listResourceFunc(ctx, &workerPods, filterLabels, &client.ListOptions{Namespace: "default"}),
 				time.Second*15, time.Millisecond*500).Should(Equal(1), fmt.Sprintf("workerGroup %v", workerPods.Items))
 		})
+
+		It("should increase replicas past maxReplicas", func() {
+			// increasing replicas to 5, which is greater than maxReplicas (4)
+			err := retryOnOldRevision(DefaultAttempts, DefaultSleepDurationInSeconds, func() error {
+				Eventually(
+					getResourceFunc(ctx, client.ObjectKey{Name: myRayCluster.Name, Namespace: "default"}, myRayCluster),
+					time.Second*9, time.Millisecond*500).Should(BeNil(), "My raycluster = %v", myRayCluster)
+				rep := new(int32)
+				*rep = 5
+				myRayCluster.Spec.WorkerGroupSpecs[0].Replicas = rep
+
+				// Operator may update revision after we get cluster earlier. Update may result in 409 conflict error.
```

Comment:
What is 409 conflict error?

Reply:
The comment is copy-pasted from a previous block of logic in the same test.

The question is significant, though.
Kubernetes uses optimistic concurrency control (backed by etcd's optimistic concurrency control).
Every K8s object carries a metadata.resourceVersion, which is a string keeping track of the history the object's kind -- each time there's an update to an object of the kind, the resourceVersion is updated.
If you read a K8s object, update a field, and attempt to reapply the object, your change may be rejected if the resource version is old. The error code for the rejected request would be 409 conflict.
For this reason, it's common in K8s code to retry on 409 errors when attempting to update a resource.

The comment is explaining the usage of the function `retryOnOldRevision` which implements the retry logic.

Reply:
An alternative to avoid 409s is to submit patches that don't include metadata.resourceVersion.
Then the latest patch will win (no concurrency control).

Reply:
Cool. Thank you for the detailed explanation!